### ISSUE TITLE
tf.nn.relu6 under CPU XLA drops NaN and returns 0, while eager/autocluster propagate NaN

### ISSUE THREAD

**Issue by MRiffiAslett** [tensorflow/tensorflow#119628]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.20.0-dev0+selfbuilt

### Custom code

Yes

### OS platform and distribution

Linux-6.14.0-37-generic-x86_64-with-glibc2.39

### Mobile device

N/A

### Python version

3.12.3

### Bazel version

bazel 9.1.0, invoked via Bazelisk

### GCC/compiler version

gcc (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0

### CUDA/cuDNN version

No CUDA/cuDNN. CPU-only environment.

### GPU model and memory

No CUDA/cuDNN. CPU-only environment.

### Current behavior?

`tf.nn.relu6` behaves differently under eager execution and CPU XLA when the input contains NaNs.

In eager execution, the NaN values are propagated:

```
eager : [nan nan nan nan]
```

However, when the same computation is compiled with XLA using `@tf.function(jit_compile=True)`, the output becomes:

```
xla   : [0. 0. 0. 0.]
```

Autocluster does not reproduce the mismatch in this run:

```
ac    : [nan nan nan nan]
```

I expected CPU XLA to match TensorFlow eager semantics and preserve the NaN values:

```
xla   : [nan nan nan nan]
```

Instead, CPU XLA appears to silently clamp/drop the NaNs and return zeros. This changes the numerical semantics of `tf.nn.relu6` and can mask invalid floating-point values in downstream computations.

This may be related to existing CPU XLA NaN-handling issues for `maximum` / `minimum`, since `tf.nn.relu6(x)` is equivalent to clamping the input into the range `[0, 6]`.

### Standalone code to reproduce the issue

```shell
#!/usr/bin/env python3
import os

os.environ["TF_XLA_FLAGS"] = "--tf_xla_auto_jit=2"

import tensorflow as tf

print("tf    :", tf.__version__)

# log(-x^2) is always NaN for positive finite x.
# This avoids using tf.constant(float("nan")) directly.
x = tf.constant([4.0, 5.0, 6.0, 7.0], dtype=tf.float32)
nan_input = tf.math.log(tf.math.negative(tf.math.square(x)))

print("input :", nan_input.numpy())
print("eager :", tf.nn.relu6(nan_input).numpy())

@tf.function(jit_compile=True)
def f_xla(t):
    return tf.nn.relu6(
        tf.math.log(tf.math.negative(tf.math.square(t)))
    )

print("xla   :", f_xla(x).numpy())

@tf.function
def f_ac(t):
    return tf.nn.relu6(
        tf.math.log(tf.math.negative(tf.math.square(t)))
    )

print("ac    :", f_ac(x).numpy())
```

### Relevant log output

```shell
tf    : 2.20.0-dev0+selfbuilt
input : [nan nan nan nan]
eager : [nan nan nan nan]
xla   : [0. 0. 0. 0.]
ac    : [nan nan nan nan]
```

**Comment by bharatjetti** [comment]
Hi @MRiffiAslett , Thanks for reporting this and sharing the details.
We’ve received the issue and are taking a look. We’ll follow up here once we have more information or if we need any additional inputs.
Appreciate your patience.

**Comment by Kayyuri** [comment]
Hi @MRiffiAslett, Thank you for your patience, and apologies for the late reply.

I tested this with the latest version of `TensorFlow 2.21`, and the outputs are consistent across eager, XLA, and autocluster execution. The NaN values are propagated in all cases `[nan nan nan nan]`. Kindly refer to this [gist](https://colab.sandbox.google.com/gist/Kayyuri/fc36c4c2090f4b032b674b450f1bf642/119628.ipynb) file.
Could you please try reproducing it with the latest version of  TensorFlow and let us know if you still encounter any issue?

Thanks!

**Comment by MRiffiAslett** [comment]
Thanks for testing this on 2.21, @Kayyuri, good to hear it's resolved there. I don't have TF 2.21 or tf-nightly readily available in my current test environment (I'm on a self-built 2.20.0-dev0 snapshot from August), so I can't independently pin down the exact fix commit, but the symptom lines up well with XLA's xla_cpu_enable_fast_min_max flag: relu6 lowers to Clamp(0, x, 6), which routes through EmitFloatMax/EmitFloatMin, and a fast min/max defaulting to true would silently drop NaN exactly as observed. If that default (or Clamp's NaN handling directly) changed between my snapshot and 2.21, that would account for it.
